DIPs/DIP-1NN4-QFS.md

```diff
@@ -0,0 +1,573 @@
+1. [Discussion: `opApply` with Type Inference and Templates?](https://forum.dlang.org/post/zkovjshfktznepertjay@forum.dlang.org)
+1. [Discussion: Parameterized delegate attributes](https://forum.dlang.org/post/ovitindvwuxkmbxufzvi@forum.dlang.org)
+1. [Discussion: `@nogc` with `opApply`](https://forum.dlang.org/thread/erznqknpyxzxqivawnix@forum.dlang.org)
+1. [Issue: `opApply` and nothrow don't play along](https://issues.dlang.org/show_bug.cgi?id=14196)
+
+## Contents
+
+* [Terms and Definitions](#terms-and-definitions)
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+  * [Attribute Checking inside Functionals](#Attribute-Checking-inside-Functionals)
+  * [Attribute Inference for Functional Templates](#Attribute-Inference-for-Functional-Templates)
+  * [Attribute Checking inside Contexts](#Attribute-Checking-inside-Contexts)
+* [Examples](#examples)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Terms and Definitions
+
+The attributes `pure`, `nothrow`, `@safe`, and `@nogc` will be called *warrant attributes* in this DIP.
+Notably absent are `@system` and `@trusted` as they do not warrant any compiler-side checks.
+
+A type is called *essentially `T`* if it is
+* a possibly qualified version of `T` — or
+* a possibly qualified pointer to essentially `T` — or
+* a possibly qualified slice of essentially `T` — or
+* a possibly qualified static array of essentially `T` — or
+* a possibly qualified associative array with value type an essentially `T`.
+
+In this DIP, *FP/D (type)* will be an abbreviation for *function pointer (type) or delegate (type)*.
+Also, *eFP/D (type)* will be used for *essentially an FP/D (type)*.
+
+An *essential call* of an eFP/D means an expression that is: In case of the eFP/D being
+* a possibly qualified FP/D: A call (in the regular sense) to that object.
+* a pointer `p` to an eFP/D: An essential call to `*p`.
+* a slice or a static or associative array `arr` to an eFP/D: An essential call to `arr[i]` for a suitable index `i`.
+
+This document makes use of the terms *(function) parameter* and *(function) argument* in a very precise manner.
+As a reminder for the reader, a parameter is a variable declared at a specific place,
+namely the function signature;
+an argument is an expression on the uppermost level in a function call.
+
+A *higher-order function*, or *functional* for short, is anything that can be called
+(including any kind of functions, function pointers, delegates, and `opCall`)
+that takes one or more eFP/D types as arguments.
+
+When a higher-order function is called, there are three notable entities to commonly refer to:
+* The *context function,* or *context* for short, is the function that contains the call expresion.
+* The *functional* is the higher-order function that is called.
```

Comment:
It's highly confusing that half of the DIP uses "higher-order" while the other use "functional".
Personally I'd rather have "higher-order" used everywhere.

Reply:
I used *functional* far more often. It's easier to type and less florid. I used *higher-order function* where I considered florid language appropriate.
I'll think about it, thanks for pointing it out.